File: ts/src/rsa4096/primesearch.ts
/**
 * Ties the candidate/trial-division/Miller-Rabin pieces together into a
 * full prime search, then the FIPS 186-5 pairwise safety checks. Direct
 * port of RSA4096/primesearch.go -- see that file's doc comments for the
 * reasoning behind the single-prime-vs-pairwise check split (a review
 * finding on the Go side, ported here already fixed).
 *
 * Copyright (C) 2026 AncientHoldings GmbH. All rights reserved.
 */

import type { Blake3XofStream } from '../dalos-blake3/index.js';
import { gcd } from './bigint-math.js';
import { CANDIDATE_BITS, generateCandidate } from './candidate.js';
import { isProbablyPrime } from './millerrabin.js';
import { passesTrialDivision } from './primes.js';
import {
  type ProgressCallback,
  type ProgressStage,
  STAGE_SEARCHING_P,
  STAGE_SEARCHING_Q,
  reportProgress,
} from './progress.js';

/** Fixed, standard RSA public exponent e = 65537. Matches RSA4096/primesearch.go. */
export const PUBLIC_EXPONENT = 65537n;

export interface FindPrimeResult {
  prime: bigint;
  attempts: number;
}

/**
 * Repeatedly draws candidates from `stream`, filters through trial
 * division, subjects survivors to Miller-Rabin, and checks
 * gcd(e, candidate-1) == 1 -- a single-prime property, checked here (not
 * in the pairwise check) so a failure only costs redrawing this one
 * candidate. Matches RSA4096/primesearch.go's FindPrime exactly.
 *
 * `stage` and `onProgress` are purely observational (see progress.ts):
 * `onProgress`, if provided, is called once per candidate draw with the
 * current attempt count and a probabilistic completion estimate. Omitting
 * it (every internal caller in this package's own tests does) skips the
 * estimate computation entirely and reproduces byte-for-byte the exact
 * same `{prime, attempts}` as before progress reporting existed.
 */
export function findPrime(
  stream: Blake3XofStream,
  stage: ProgressStage,
  onProgress?: ProgressCallback,
): FindPrimeResult {
  let attempts = 0;
  for (;;) {
    attempts++;
    reportProgress(onProgress, stage, attempts);

    const candidate = generateCandidate(stream);

    if (!passesTrialDivision(candidate)) {
      continue;
    }

    if (!isProbablyPrime(candidate, stream)) {
      continue;
    }

    const candidateMinus1 = candidate - 1n;
    if (gcd(PUBLIC_EXPONENT, candidateMinus1) !== 1n) {
      // gcd(e, candidate-1) != 1: astronomically rare (~1/65537 of
      // primes); draw a fresh candidate, don't touch the OTHER prime's
      // search.
      continue;
    }

    return { prime: candidate, attempts };
  }
}

/**
 * FIPS 186-5's |p-q| > 2^(nlen/2 - 100) bound: nlen = 4096 (the RSA
 * modulus size), nlen/2 = 2048 = CANDIDATE_BITS (the size of each prime),
 * so the threshold is 2^(2048-100) = 2^1948. Matches
 * RSA4096/primesearch.go's minPrimeDistanceExponent.
 */
const MIN_PRIME_DISTANCE_EXPONENT = BigInt(CANDIDATE_BITS - 100);

/**
 * Runs the genuinely PAIRWISE safety checks (p != q, |p-q| large enough)
 * against a candidate (p, q) pair. Matches RSA4096/primesearch.go's
 * checkAuxiliaryConstraints exactly (gcd(e, prime-1) is a single-prime
 * property and lives in findPrime instead).
 */
export function passesAuxiliaryConstraints(p: bigint, q: bigint): boolean {
  if (p === q) {
    return false;
  }
  const diff = p > q ? p - q : q - p;
  const threshold = 1n << MIN_PRIME_DISTANCE_EXPONENT;
  return diff > threshold;
}

export interface FindTwoPrimesResult {
  p: bigint;
  q: bigint;
  pAttempts: number;
  qAttempts: number;
}

/**
 * Runs the full search on a single continuous stream: find p, then
 * (without resetting anything) keep searching from wherever the stream is
 * to find q, then validate the pair. On a pairwise-check failure, keeps
 * `p` and redraws only `q` -- matches RSA4096/primesearch.go's
 * FindTwoPrimes exactly.
 */
export function findTwoPrimes(
  stream: Blake3XofStream,
  onProgress?: ProgressCallback,
): FindTwoPrimesResult {
  const { prime: p, attempts: pAttempts } = findPrime(stream, STAGE_SEARCHING_P, onProgress);

  for (;;) {
    const { prime: q, attempts: qAttempts } = findPrime(stream, STAGE_SEARCHING_Q, onProgress);
    if (passesAuxiliaryConstraints(p, q)) {
      return { p, q, pAttempts, qAttempts };
    }
    // Pairwise check failed (in practice, only the |p-q| distance check
    // can fire now that gcd(e, prime-1) is checked per-prime -- p===q is
    // essentially impossible, and a too-small |p-q| is very unlikely for
    // independently-drawn 2048-bit primes but not provably impossible).
    // Keep p, draw a fresh q from further along the same stream.
  }
}

// --- Async variants: for a UI progress bar that must actually repaint ---
//
// Progress reporting alone (findPrime/findTwoPrimes above) is NOT enough
// to drive a smoothly-updating browser progress bar: RSA-4096 generation
// is CPU-bound native-BigInt arithmetic, and JavaScript's single-threaded
// event loop cannot repaint the DOM while a long synchronous function is
// still running, no matter how many times it calls a callback internally.
// The callback fires, but the browser can't draw the frame until control
// returns to the event loop.
//
// findPrimeAsync mirrors findPrime's body byte-for-byte, with exactly one
// addition: `if ((attempts & 0x07) === 0) await yieldToEventLoop();` --
// the same "yield every 8 iterations, triggered only by a public,
// data-independent counter" pattern already established in this repo for
// scalarMultiplierAsync/schnorrSignAsync (see gen1/scalar-mult.ts). The
// two bodies are kept as parallel, near-identical functions rather than
// unified into one generic yielding abstraction -- deliberately, so
// either can be read start-to-finish and audited against the other by
// direct comparison, matching this codebase's existing convention.

/**
 * Yields control to the event loop via `setImmediate` (Node) or
 * `setTimeout(_, 0)` (browser/Deno fallback) -- identical mechanism to
 * gen1/scalar-mult.ts's module-private helper of the same name (that one
 * is not exported, so it can't be imported and shared; this is a second,
 * independent copy of the same three lines, not a meaningful duplication
 * risk since there is nothing to keep "in sync" -- it's a fixed platform
 * API call).
 */
function yieldToEventLoop(): Promise<void> {
  return new Promise((resolve) => {
    const g = globalThis as { setImmediate?: (cb: () => void) => void };
    if (typeof g.setImmediate === 'function') {
      g.setImmediate(resolve);
    } else {
      setTimeout(resolve, 0);
    }
  });
}

/**
 * Async mirror of findPrime -- same search, same result, but yields to
 * the event loop every 8 candidate draws so a UI progress bar can
 * actually repaint during the search. See the section comment above for
 * why this is a separate function rather than a flag on findPrime.
 */
export async function findPrimeAsync(
  stream: Blake3XofStream,
  stage: ProgressStage,
  onProgress?: ProgressCallback,
): Promise<FindPrimeResult> {
  let attempts = 0;
  for (;;) {
    attempts++;
    reportProgress(onProgress, stage, attempts);
    if ((attempts & 0x07) === 0) {
      await yieldToEventLoop();
    }

    const candidate = generateCandidate(stream);

    if (!passesTrialDivision(candidate)) {
      continue;
    }

    if (!isProbablyPrime(candidate, stream)) {
      continue;
    }

    const candidateMinus1 = candidate - 1n;
    if (gcd(PUBLIC_EXPONENT, candidateMinus1) !== 1n) {
      continue;
    }

    return { prime: candidate, attempts };
  }
}

/** Async mirror of findTwoPrimes, built on findPrimeAsync. */
export async function findTwoPrimesAsync(
  stream: Blake3XofStream,
  onProgress?: ProgressCallback,
): Promise<FindTwoPrimesResult> {
  const { prime: p, attempts: pAttempts } = await findPrimeAsync(
    stream,
    STAGE_SEARCHING_P,
    onProgress,
  );

  for (;;) {
    const { prime: q, attempts: qAttempts } = await findPrimeAsync(
      stream,
      STAGE_SEARCHING_Q,
      onProgress,
    );
    if (passesAuxiliaryConstraints(p, q)) {
      return { p, q, pAttempts, qAttempts };
    }
  }
}

File: research/rsa4096-poc/primesearch.go
// This file ties steps 2-5 together: the guess-and-check loop that finds one
// 2048-bit prime (steps 2-4), then the pairwise safety checks FIPS 186-5
// requires once both primes are found (step 5). See
// /.docs/deterministic-rsa4096-from-seed.md §5.2-§5.2.7.
package main

import (
	"errors"
	"io"
	"math/big"
)

// publicExponent is the fixed, standard RSA public exponent e = 65537.
// Never derived from the seed or the primes -- see the chat log for why
// this specific number is universal practice.
var publicExponent = big.NewInt(65537)

// FindPrime repeatedly draws candidates from stream (step 2), filters them
// through trial division (step 3), and subjects survivors to Miller-Rabin
// (step 4), continuing on the SAME stream without ever rewinding or
// reseeding, until one passes -- AND satisfies gcd(e, candidate-1) == 1.
// That last check is a single-prime property (not pairwise), so it belongs
// here rather than in checkAuxiliaryConstraints: if it fails, we only need
// to throw away this one candidate and keep searching, not discard an
// already-found, already-valid partner prime too. (An earlier version of
// this code checked gcd(e, p-1) and gcd(e, q-1) only after BOTH primes had
// already been found, in the pairwise check -- which meant a ~1-in-65537
// single-prime failure wasted a perfectly good already-verified partner
// prime and its full Miller-Rabin search. Caught in review before this ever
// shipped anywhere; fixed here.)
//
// Returns the prime found and the number of candidates it took (useful for
// sanity-checking against the expected ~710-draw average from the density
// math).
func FindPrime(stream io.Reader) (*big.Int, int, error) {
	attempts := 0
	for {
		attempts++
		candidate, err := GenerateCandidate(stream)
		if err != nil {
			return nil, attempts, err
		}

		if !PassesTrialDivision(candidate) {
			continue
		}

		isPrime, err := IsProbablyPrime(candidate, stream)
		if err != nil {
			return nil, attempts, err
		}
		if !isPrime {
			continue
		}

		candidateMinus1 := new(big.Int).Sub(candidate, bigOne)
		if new(big.Int).GCD(nil, nil, publicExponent, candidateMinus1).Cmp(bigOne) != 0 {
			// gcd(e, candidate-1) != 1: this specific prime can't be used
			// with our fixed e (AssembleKey would have no inverse for it).
			// Astronomically rare (~1/65537 of primes), but when it does
			// happen the fix is to draw a fresh candidate, not to touch
			// whatever the OTHER prime search found.
			continue
		}

		return candidate, attempts, nil
	}
}

// minPrimeDistanceExponent implements the FIPS 186-5 auxiliary requirement
// that |p-q| be large enough to resist Fermat factorization: the standard
// bound is |p-q| > 2^(nlen/2 - 100), where nlen is the RSA modulus size (4096
// here), so nlen/2 = 2048 (the size of each prime) and the threshold is
// 2^(2048-100) = 2^1948.
const minPrimeDistanceExponent = candidateBits - 100

// checkAuxiliaryConstraints runs step 5's genuinely PAIRWISE safety checks
// against a candidate (p, q) pair -- properties that depend on both primes
// together, not on either one alone. (gcd(e, prime-1) is a single-prime
// property and lives in FindPrime instead -- see its comment for why.)
// Returns a non-nil error naming which check failed if any do; the caller's
// job is to discard BOTH primes and re-search, since a pairwise failure
// isn't attributable to just one of them.
func checkAuxiliaryConstraints(p, q *big.Int) error {
	if p.Cmp(q) == 0 {
		return errors.New("auxiliary check failed: p == q")
	}

	diff := new(big.Int).Sub(p, q)
	diff.Abs(diff)
	threshold := new(big.Int).Lsh(bigOne, minPrimeDistanceExponent)
	if diff.Cmp(threshold) <= 0 {
		return errors.New("auxiliary check failed: |p-q| too small (Fermat-factorization risk)")
	}

	return nil
}

// FindTwoPrimes runs the full step 2-5 pipeline on a single continuous
// stream: find p, then (without resetting anything) keep searching from
// wherever the stream is to find q, then validate the pair. If the pair
// fails an auxiliary check, BOTH primes are discarded and the search
// continues from the current stream position for a fresh pair -- the
// stream is never rewound, so every attempt (successful or not) permanently
// consumes stream material, exactly as the "structural, not statistical"
// correctness bar demands: nothing is ever reused or replayed.
func FindTwoPrimes(stream io.Reader) (p, q *big.Int, pAttempts, qAttempts int, err error) {
	for {
		p, pAttempts, err = FindPrime(stream)
		if err != nil {
			return nil, nil, 0, 0, err
		}
		q, qAttempts, err = FindPrime(stream)
		if err != nil {
			return nil, nil, 0, 0, err
		}

		if checkErr := checkAuxiliaryConstraints(p, q); checkErr == nil {
			return p, q, pAttempts, qAttempts, nil
		}
		// The pair failed a genuinely pairwise check (in practice this can
		// only be the |p-q| distance check now that gcd(e, prime-1) is
		// checked per-prime in FindPrime -- p==q is essentially impossible,
		// and a too-small |p-q| is very unlikely for independently-drawn
		// 2048-bit primes but not provably impossible). Both primes are
		// discarded together because the failure is a property of the pair,
		// not attributable to either one alone. Loop and draw a fresh pair
		// from further along the same stream.
	}
}
